### DISCUSSION TITLE
SRS Lagrange size check is wrong

### DISCUSSION THREAD

**Discussion by secure12** [Consensys/gnark#1088] [General]
https://github.com/Consensys/gnark/blob/9bb41533c9f58e4d1ee0513e13618cdf1330e36a/backend/plonk/bw6-761/setup.go#L108

Should not it be okay to use an SRS with larger size? So this should instead be:
```
if len(srsLagrange.Pk.G1) < int(domain.Cardinality) {
```

**Comment by ivokub** [comment]
Unfortunately not - we need to compute `(x-alpha^0)(x-alpha)...(x-alpha^n)` where `alpha` is generator. If `alpha` is exactly `n`-th root of unity, then it can be given as `(x^n-1)`. Due to this the Lagrange generator depends on the size of the circuit.

**Reply by secure12** [reply]
OK. Then perhaps the error message is a bit misleading...

**Reply by secure12** [reply]
If the SRS Lagrange must be the same as the circuit size, how would it be generated in advance?

**Reply by secure12** [reply]
Nevermind. Apparently `gnark-crypto offers the ToLagrangeG1 method to convert it to lagrange form.`

But I think at least the error message could be improved.

**Reply by ivokub** [reply]
Have a look at the PLONK recursive verifier tests: https://github.com/Consensys/gnark/blob/master/std/recursion/plonk/verifier_test.go#L347-L373

There, we generate the SRS from maximum circuit size and then compute per domain Lagrange SRS.

And I do agree that the error message should be more descriptive. We actually have a related open issue for that: https://github.com/Consensys/gnark/issues/864

**Reply by secure12** [reply]
The example code is helpful. Thank you!
